### PR TITLE
MRG: Allow filtering Evoked and Epochs, add pad arg

### LINKED ISSUE #4314: filtering evoked and epochs
Hi,
there is no facility for filtering Epochs or Evoked, except by Savitzky-Golay polynomial fit. Apparently this is to encourage filtering at raw data level instead (to avoid edge artifacts). Couple of points/questions:

- Why Savitzky-Golay? As Ronald Schafer has pointed out [1], S-G is equivalent to convolution with a finite impulse response, but actually has properties inferior to standard FIR design methods.
- Why are edge artifacts so important? If significant ones occur with a given filtering method, they are limited in duration and in most cases easily ignored (IMO).

Filtering raw data is very slow in many cases, so it would be nice to have .filter() method to quickly play around with frequency domain properties of Evoked and Epochs. For example, the Elekta software allows this. Of course one has to be careful not to distort data with e.g. carelessly applied highpass filters, but the same applies equally to filtering Raw.

[1] https://ai.berkeley.edu/~ee123/sp15/docs/SGFilter.pdf

### PR DESCRIPTION
**Pull request by larsoner**
Closes #4314.

@jjnurminen can you see if this is the behavior you wanted / would expect?

### CODE REVIEW

**Comment by jjnurminen** [comment]
Wow that was fast.
I'll take a look ASAP.

**Comment by larsoner** [comment]
Latest commit should fix CIs, ready for review/merge from my end

File: mne/filter.py [hunk 1]
```diff
@@ -1899,10 +1922,211 @@ def savgol_filter(self, h_freq, copy=False):
...
+            time-domain design technique that generally gives improved
+            attenuation using fewer samples than "firwin2".
+        skip_by_annotation : str | list of str
+            If a string (or list of str), any annotation segment that begins
+            with the given string will not be included in filtering, and
+            segments on either side of the given excluded annotated segment
+            will be filtered separately (i.e., as independent signals).
+            The default in 0.16 (``'edge'``) will separately filter any
+            segments that were concatenated by :func:`mne.concatenate_raws`
+            or :meth:`mne.io.Raw.append`. To disable, provide an empty list.
```
Comment (agramfort) [inline mne/filter.py]: does not apply here.

Comment (agramfort) [reply mne/filter.py]: or at least on Epochs or Evoked

Comment (larsoner) [reply mne/filter.py]: good catch, fixed

**Comment by agramfort** [comment]
@larsoner can you show us how savgol filter and the new filter compares for signal as long as a typical epoch when using same l_freq and h_freq?

**Comment by larsoner** [comment]
Sure I can maybe update the tutorial

**Comment by agramfort** [comment]
Or just show it here first ?

Alex

**Comment by larsoner** [comment]
Example, filter at 100 Hz:
```
import numpy as np
import mne
import matplotlib.pyplot as plt

sfreq = 1024.
data = np.random.RandomState(0).randn(60, 60, 1024)
info = mne.create_info(data.shape[1], sfreq, 'eeg')
epochs = mne.EpochsArray(data, info)

epochs_filt_shallow = epochs.copy().filter(
    None, 100., h_trans_bandwidth=300, fir_design='firwin',
    filter_length=11)
epochs_filt_medium = epochs.copy().filter(
    None, 100., h_trans_bandwidth=150, fir_design='firwin',
    filter_length=23)
epochs_filt_auto = epochs.copy().filter(None, 100., fir_design='firwin')
epochs_sav = epochs.copy().savgol_filter(100., verbose=True)

for epochs, label in (
        (epochs_filt_auto, 'FIR auto'),
        (epochs_sav, 'Sav-Gol (11 points)'),
        (epochs_filt_shallow, 'FIR shallow (11 points)'),
        (epochs_filt_medium, 'FIR medium (23 points)'),
        ):
    epochs = epochs.get_data()
    epochs.shape = (60, -1)
    psd, freqs = mne.time_frequency.psd_array_welch(
        epochs, sfreq, n_fft=1024, n_overlap=0)
    plt.plot(freqs, 20 * np.log10(np.mean(psd, axis=0)), label=label)
plt.legend()
plt.xlabel('Frequencies (Hz)')
plt.ylim([-120, -50])
```
Our implementation of Sav-Gol by default uses only 11 points, which has a huge transition bandwidth, and huge stop-band ripple:




Maybe we could make Sav-Gol. But in any case, the FIR code seems to be working just fine.

**Comment by agramfort** [comment]
how about the signal waveforms eg on sample on somato?

also when you use classical 0-40 or 1-40 bandpass?

**Comment by larsoner** [comment]
Demo script for a reasonable 40 Hz lowpass:
```
import mne
fname = mne.datasets.sample.data_path() + '/MEG/sample/sample_audvis_raw.fif'
raw = mne.io.read_raw_fif(fname, preload=True)
events = mne.find_events(raw, 'STI 014')
raw.pick_types(meg='mag')

kwargs = dict(l_freq=None, h_freq=40., fir_design='firwin')
raw_filt = raw.copy().filter(**kwargs)

epochs_prefilt = mne.Epochs(raw_filt, events)
epochs_postfilt = mne.Epochs(raw, events, preload=True).filter(**kwargs)

evoked_prefilt = epochs_prefilt.average()
evoked_postfilt = epochs_postfilt.average()

evoked_prefilt.plot()
evoked_postfilt.plot()
```
40 Hz lowpass is indistinguishable on `sample` data:
raw
epochs

`epochs.plot()` look almost identical, too.

High-passing is not a good idea (filter is way too long for e.g. 1 sec epochs), and it warns about this:
```
filter_length (1983) is longer than the signal (421), distortion is likely. Reduce filter length or filter a longer signal.
```
Some mild distortion is indeed observed when doing `l_freq=1., h_freq=40.`:
raw_hp
epochs_hp

**Comment by jjnurminen** [comment]
This is really cool and will make life easier when playing with Evokeds.

When using the default FIR design, I get a lot of warnings :

```
In [3]: evf = evoked.filter(0, 60, method='fir')
Setting up low-pass filter at 60 Hz
fir_design defaults to "firwin2" in 0.15 but will change to "firwin" in 0.16, set it explicitly to avoid this warning.
h_trans_bandwidth chosen to be 15.0 Hz
Filter length of 1320 samples (0.440 sec) selected
filter_length (1321) is longer than the signal (1051), distortion is likely. Reduce filter length or filter a longer signal.
```

The filter seems pretty long though it's just a 60 Hz lowpass, are the design criteria maybe a bit too strict? Off the top of my head, I cannot remember how problematic it is to use overlong FIR filters (longer than signal).

**Comment by larsoner** [comment]
Filter length depends on 1) sample rate, 2) transition bandwidth, and 3) window function. A 15 Hz transition band with `firwin2` (not as good as `firwin`!) design and `hamming` window should require 1321 samples if your sampling rate is `1321 * 15 / (3.3 * 2)` which is ~3000 Hz, is it? If so this all sounds reasonable. But you should really heed the warning and use `firwin` because it both uses a shorter filter (which should make the other warning go away) length and has in general better characteristics.

**Comment by larsoner** [comment]
(Sorry, I should add it depends also on 4) design method, as firwin2 will be twice as long as firwin)

**Comment by agramfort** [comment]
@larsoner please rebase, make CIs green and +1 for MRG

**Comment by larsoner** [comment]
@jjnurminen okay for you?

**Comment by jjnurminen** [comment]
So far I had no problems with this one, so +1 from me.

**Comment by agramfort** [comment]
thx @larsoner !

**Pull request merged**